### ISSUE TITLE
Secret Contract Code Verification System

### ISSUE THREAD

**Issue by Gago55** [SecretFoundation/Grants#7]
## Project Description

We intend to create **Secret Contract code verification system** like the Ethereum ecosystem has, particularly etherscan.io.

## Problem / Solution
**Problem**

We see two problems that intersect:

1. There is no place where developers can relate* the source code to Contract address and verify it.
2. For Secret Contract users at this point it is very hard to find** the source code of the contract and be convinced that the contract is deployed from it.

\*We know that in contract deploying process the developer can provide a link to the source, that will be written in the blockchain. **But** a bad actor can put any link that can create confusion for the user.

\**At the investigation time for this offer we had a hard time to find the most executed contract in our network which is SSCRT. Because the repo has been updated, it wasn't clear which commit had been deployed. Fortunately, we have active admins in dev-discussion channel in discord who pointed to the commit of the deployed source.

**Solution**

* Docker image that will verify contract code.

* Open source web app  that will:
  * Show contracts list, with info about verification status. 
  * Show contract source code if it is verified.
  * Give opportunity to provide a source code in order to verify.

## Detailed product description

**Docker - Secret Contract Verifier**

The Docker will take 3 inputs:
* Source code of the contract that the user wants to verify.
* Secret Contract Optimizer version.
* Hash of Deployed Secret Contract.

And give a boolean output about the success of verification.

**Web App**
We will create an open-source web app using MERN/Typescript.

In main page will be listed all deployed contracts in Mainnet with next columns:

* Contract **Label**
* Contract **CodeID**
* Contract **Verification Status**
* Show/Provide source code Button

On click in a contract row will be directed to detailed view which will contain:

* General contract Information
* The source code of contract in case it was verified previously.
* "Provide source code" button in case it wasn't verified previously.

On click "Provide source code" button, will be opened dialog where developer must choose the version of Secret Contract Optimizer that he/she used, and upload zip file of contract project. After completion of the uploading process, the app will give a temp link. If a developer opens a link, he/she can see the logs of the verification process. If the result of the Secret Contract Verifier is "TRUE", the status of that contract will be changed to "Verified". Also, all contracts with the same CodeID will get "Verified" status. Source code of each "Verified" contract will be saved in DB in order to serve to users.

So Why Docker Image? In case users think that we (the web app) can be bad actor, they can verify the contract by themself easily.

## Value capture for Secret Network ecosystem

Our idea is that the system finally will be deployed on different explorers, specially in one that will be baked by Secret Foundation. And We will have more strong and **trustless** ecosystem.

Also, the verification process of Secret Contract will be easier with Secret Contract Verifier (docker) for users.

## Team members
* Gagik Khachatyan *(21  y.o.)*
* Manvel Arzumanyan *(41  y.o.)*

## Team's experience
* Gagik is a student of computer science - Information Security (NPUA). He has 3 years of experience in developing with different techs and currently landed to MERN combination using Typescript. He is involved in Secret Network, thanks to tutorials created by Figment. He is actually **Secret Agent**. 
* Manvel was in Secret Network since September of 2017 from Enigma's times. He participated in ICO without luck. Privacy enthusiast. Promoter of Secret Network without having titles. In the 3D world since 2000, developing 3D in web since 2013. 

We've been a team for 5 years. Spent major time on proprietary 3D Configurator software using WebGL.

## Team Code Repos
* https://github.com/gago55
* https://github.com/manvelarz

## Team LinkedIn Profiles
https://www.linkedin.com/in/manvel-arzumanyan-18623b5a/

## Development Roadmap
The problem was seen from the start of the Mainnet. We were waiting for the solution to come on its own (no luck). When we saw the Grant Program announcement we decided to go forward with it. **So** the Roadmap was started a month ago. We have spent a lot of time to understand all aspects of the problem deeply and to plan the solution.

We will require 25-40 days to complete this project. We intend to have one developer full-time and 1 part-time. Also, we will provide infrastructure for web app and support it for one year. At a total cost of $24,000.
* first we'll create the Verifier Docker image - 5-10 days 
* web app will be finished in about 30 days

Ideally, we can receive payments in 2 disbursements, 25% at the beginning of the grant and 75% when the development work is completed.

We would like to receive 100 % in SCRTs, as we believe in the success of our Network.

## Additional Information

We are open to estimate in the future, the version of the Verifier (docker) that will verify the Secret Contract that has been created without Secret-Contract-Optimizer Docker.

We believe that Secret Network will be a big part of the privacy ecosystem for our civilization.
We’ll be proud to be part of it.

**Comment by assafmo** [comment]
Hi @gago55 and @manvelarz! Thanks for this submission.

This feature is indeed something I personally wanted since before mainnet, but both explorers did not have the capacity to implement it.

I believe that an additional web app is not needed here, and this should be integrated into an explorer. In the blockchain itself everything is ready for this (`builder` and `source` fields), and it's now a matter of implementing reproducible builds for contracts (which might be not that easy [[1]](https://users.rust-lang.org/t/testing-out-reproducible-builds/9758)[[2]](https://www.reddit.com/r/rust/comments/jct0y4/when_reproducible_builds/)[[3]](https://github.com/rust-lang/rust/labels/A-reproducibility)) and setting standards for contract builders of how to use the `builder` and `source` fields when uploading a new contract to the blockchain.

So to summarize my thoughts:
- Instead of the web app, I'd like to see an integration plan with at least https://secretnodes.com (Cc @moonstash @jacobtjackson). Bonus points for @Cashmaney.
- I'd like to see a plan for reproducible builds, because right now I doubt it'll take 5-10 days to achieve this.

Again, thank you for wanting to improve Secret Network!

**Comment by manvelarz** [comment]
@assafmo , @reuvenpo 
>Instead of the web app, I'd like to see an integration plan with at least https://secretnodes.com

Our intend is to create an **open source** app, so any interested one can take and use it. Maybe someone will upgrade it and keep the source closed or better, contribute his source. In any single case Secret Network will benefit incrementally. So The **OPEN SOURCE** ideology will start to benefit to our network and users too. 
Our vision is that the web app will be the foundation for feature open source explorer for Secret Network. Ideally it will be at [https://explorer.scrt.network/](https://explorer.scrt.network/). We are planning to propose the Open source explorer after finishing this project. The benefit in doing in this sequence is that other explorers can start to integrate this functionality meanwhile we will be working in creation of open source explorer. 

>I'd like to see a plan for reproducible builds, because right now I doubt it'll take 5-10 days to achieve this.

I don't understand your point. Do you mean  **reproducible builds** without the Secret-Contract-Optimizer docker?
Because with that docker in our understanding Contracts **are** reproducible. We've tried and succeeded(achieved blockchain's Hashes) on reproduction of several contracts, Including SSCRT(CodeID-5) and ETH-reward(CodeID-23).


************************************

Quote from [www.secret.foundation](https://www.secret.foundation/)
>The mission of Secret Foundation is to support the global adoption of **open-source**, privacy-centric technologies that help protect and empower users around the world.

**Comment by moonstash** [comment]
> @assafmo , @reuvenpo 
> 
> >Instead of the web app, I'd like to see an integration plan with at least https://secretnodes.com
> 
> 
> 
> Our intend is to create an **open source** app, so any interested one can take and use it. Maybe someone will upgrade it and keep the source closed or better, contribute his source. In any single case Secret Network will benefit incrementally. So The **OPEN SOURCE** ideology will start to benefit to our network and users too. 
> 
> Our vision is that the web app will be the foundation for feature open source explorer for Secret Network. Ideally it will be at [https://explorer.scrt.network/](https://explorer.scrt.network/). We are planning to propose the Open source explorer after finishing this project. The benefit in doing in this sequence is that other explorers can start to integrate this functionality meanwhile we will be working in creation of open source explorer. 
> 
> 
> 
> >I'd like to see a plan for reproducible builds, because right now I doubt it'll take 5-10 days to achieve this.
> 
> 
> 
> I don't understand your point. Do you mean  **reproducible builds** without the Secret-Contract-Optimizer docker?
> 
> Because with that docker in our understanding Contracts **are** reproducible. We've tried and succeeded(achieved blockchain's Hashes) on reproduction of several contracts, Including SSCRT(CodeID-5) and ETH-reward(CodeID-23).
> 
> 
> 
> 
> 
> ************************************
> 
> 
> 
> Quote from [www.secret.foundation](https://www.secret.foundation/)
> 
> >The mission of Secret Foundation is to support the global adoption of **open-source**, privacy-centric technologies that help protect and empower users around the world.
> 
> 

The community pool already funded an open source block explorer with 250k scrt last year. That work which is still open source, has since turned into the secretnodes.com community explorer which is fully maintained and improved upon based on a self sustainable model that allows us to not ask for additional funding in relation to the explorer. We would be happy to use our resources to incorporate the work you are proposing on the contract verifier if you're willing to work with us, otherwise the feature is already on our roadmap and we will simply do it ourselves (though we'd prefer to collaborate to the max extent there is interest for / that makes sense).

**Comment by assafmo** [comment]
> I don't understand your point. Do you mean  reproducible builds without the Secret-Contract-Optimizer docker?
Because with that docker in our understanding Contracts are reproducible. We've tried and succeeded(achieved blockchain's Hashes) on reproduction of several contracts, Including SSCRT(CodeID-5) and ETH-reward(CodeID-23).

Secret-Contract-Optimizer docker does not guarantee reproducible builds.

**Comment by reuvenpo** [comment]
To clarify, the docker _helps_ with creating reproducible builds, but the full process requires [support from the contract itself](https://github.com/enigmampc/snip20-reference-impl/blob/7429b4fa6492e6edd17540a2f0dc34efaefd59c3/Cargo.toml#L18-L27). Additionally, many of the issues described in the links Assaf posted above apply to specific architectures and specific binary formats. I understand that most of those do not apply to Wasm code that is compiled without backtraces, in a controlled environment (where paths are well known).
This work requires some experimentation to see what is possible and what poses a challenge. I am pretty sure though that if someone builds the contract locally without the docker image, reproducing the build becomes harder, but i also think that all the users have to do then is also provide information about what version of Rust was used, what build features were enabled, and what directory the build ran on in the programmer's machine.

**Comment by cankisagun** [comment]
This proposal is approved.

**Comment by Gago55** [comment]
Hello, dear Secret Network community. 
Congratulations to all of us 🎉 one more problem was solved for our network.

The proposal is completed, and we have up and running Secret Contract Verifier - http://secret-contracts.com.
The source code is in https://github.com/Gago55/Secret-Contract-Verifier.
Also, we have done an extra feature, after successful verification of Code (Secret Contract) the source will be automatically pushed to https://github.com/secretcontracts , and we will all verified Codes (Secret Contracts) in one familiar place too.

Little bit details:
As you can see the proposal was approved on May 12
We received the first disbursement on May 20 and started to work on it the next day
We worked hard to finish the project on time and done it in 40 days as we predicted

We had a demonstration meeting today with @reuvenpo @toml01
We consider the project finished. 

Thanks to all who supported us, specially  @reuvenpo and @cankisagun

**Comment by assafmo** [comment]
Hi guys,nice work!
Where are the instructions for contact developers on how to make their code verified? And is it possible to turn an unverified code to a verified one?

**Comment by manvelarz** [comment]
Hi @assafmo . I've prepared a video instruction and I hope it will respond to your second question. Please take a look.
[Video tutorial](https://www.youtube.com/watch?v=A4fabytonCg)

**Comment by moonstash** [comment]
This is awesome!😎
